### PR TITLE
Cleanup GitHub Actions workflow

### PR DESCRIPTION
**Pull request by pellared**
## What

Refactor the GH Actions workflow to use a "matrix".

See the results: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/actions/runs/1443822406

## Why

1. Less likely to make an error when "copy-pasting" steps between the jobs
1. Easier to ensure consistency
1. Less noise
1. Even currently we are not using the GH workflow generated by Nuke

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,90 +1,26 @@
...
-
 name: ci
 
 on:
   push:
-    branches:
-      - main
-      - 'refs/tags/*'
-    paths:
-      - '!docs/*'
```
Comment (zacharycmontoya) [inline .github/workflows/ci.yml:25]: Did you mean to remove the docs exclusion? It seems useful to exclude docs-only changes (unless some checks are being enforced and therefore they would still have to run)

**Review by zacharycmontoya — APPROVED** [review]
LGTM

File: .github/workflows/ci.yml (hunk 1 above)
Comment (pellared) [reply .github/workflows/ci.yml:25]: > unless some checks are being enforced 

In "legacy main" the builds were required. Now there are not, but I am not sure if this is as expected.

If we do not want to require any builds then I can bring it back.

Comment (zacharycmontoya) [reply .github/workflows/ci.yml:25]: There's no harm in leaving this change. If we feel like it slows down development, we can always add the exclusion back in later

**Review by rajkumar-rangaraj — APPROVED** [review]

**Review by nrcventura — APPROVED** [review]

**Pull request merged**